videohash.c

```diff
@@ -0,0 +1,324 @@
+    
+    // Load Image
+    magick_wand = load_image_from_frame(filename, frame_number, frame_data);
+    if(!magick_wand) return NULL;
+    
+    // Compute Image Hash
+    result = compute_image_hash(magick_wand, bits, quick, &hash);
+    
+    switch(result)
+    {
+        case 0: break;
+        
+        case 1: {
+            fprintf(stderr, "Error computing blockhash for the zero-sized frame #%llu of the video file '%s'.", 
+        	    (unsigned long long)frame_number, filename);
+            break;
+        }
+        
+        case 2: {
+            fprintf(stderr, "Error converting image data to RGBA for the frame #%llu of the video file '%s'.\n", 
+        	    (unsigned long long)frame_number, filename);
+            break;
+        }
+        default: {
+            fprintf(stderr, "Error computing blockhash for the frame #%llu of the video file '%s'.", 
+        	    (unsigned long long)frame_number, filename);
+            break;
+        }
+    }
+    
+    // Cleanup temporary data
+    DestroyMagickWand(magick_wand);
+    
+    // Report the result
+    return hash;
+}
+
+typedef struct _video_frame_info {
+    size_t frame_number;
+    int* hash;
+} video_frame_info;
+
+#define HASH_PART_COUNT 4 
+
+int process_video(char *filename, int bits, int quick, int debug)
+{
+    size_t i;
+    int result = 0;
+    CvCapture* capture;
+    size_t frame_count;
+    video_frame_info hash_frames[HASH_PART_COUNT];
+    size_t next_hash_frame;
+    size_t current_frame;
+    int* hash = NULL;
+ 
+    // Initialize data
+    memset(&hash_frames[0], 0, sizeof(hash_frames));
+    
+    // Open video file
+    capture = cvCreateFileCapture(filename);
+    if(!capture) {
+        fprintf(stderr, "Error opening video file '%s'.", filename);
+        return -1;
+    }
+    
+    // Get frame count
+    frame_count = (size_t)cvGetCaptureProperty(capture, CV_CAP_PROP_FRAME_COUNT);
+    // Determine first and last hash frames
+    if(frame_count < 11) {
+        if(frame_count > 0) {
+            hash_frames[0].frame_number = 0;
+            hash_frames[3].frame_number = frame_count - 1;	  
+        } else {
+            // Handle special zero frames case
+            result = -1;
+            fprintf(stderr, "Error computing blockhash for zero-frame video file");
+            goto cleanup;
+        }
+    } else {
+        hash_frames[0].frame_number = 10;
+        hash_frames[3].frame_number = frame_count - 11;
+    }
+    
+    // Determine middle hash frames
+    hash_frames[1].frame_number = (size_t)floor(frame_count * 0.35);
+    hash_frames[2].frame_number = (size_t)floor(frame_count * 0.7);
+    
+    // Read frames one by one and compute hash for the selected frames
+    next_hash_frame = 0;
+    for(current_frame = 0; current_frame < frame_count; ++current_frame) {
+        IplImage* frame_image = cvQueryFrame(capture);
+        if(!frame_image) {
+            // Signal an error only if we don't manage to get the last frame
```

Comment:
This seems to happen on around 30% of ogv files. This change, to only count it as an error if in fact we don't manage to catch all the frames we need seem to work well, but still, we get various errors on about 10% of the files when there's simply not as many frames as would be expected. I suspect that this is because frame numbers are somewhat imprecise: the calculation of available frame numbers from the header may be calculated using fps and timestamps. While reading through the file, it may turn out that the number of actual frames deviated from this.